autolink/postlink/applicationLinker.js

```diff
@@ -1,68 +1,126 @@
+
+    if (this.navigationApplicationSuccess && this.navigationHostSuccess && this.soLoaderInitSuccess) {
+      infon('MainApplication.java linked successfully!\n')
+    } else if (!this.navigationApplicationSuccess && !this.navigationHostSuccess && !this.soLoaderInitSuccess) {
+      errorn(
+        'MainApplication.java was not successfully linked! Please check the information above:\n   https://wix.github.io/react-native-navigation/docs/installing#3-update-mainapplicationjava'
+      )
+    } else {
+      warnn(
+        'MainApplication.java was partially linked. Please check the information above and complete the missing steps manually:\n   https://wix.github.io/react-native-navigation/docs/installing#3-update-mainapplicationjava'
+      )
     }
   }
 
   _extendNavigationApplication(applicationContent) {
     if (this._doesExtendApplication(applicationContent)) {
-      debugn("   Extending NavigationApplication");
+      debugn('   Extending NavigationApplication')
+      return applicationContent
+        .replace(/extends\s+Application\s+implements\s+ReactApplication/gi, 'extends NavigationApplication')
+        .replace(
+          'import com.facebook.react.ReactApplication;',
+          'import com.reactnativenavigation.NavigationApplication;'
+        )
+    }
+
+    if (this._hasAlreadyLinkedApplication(applicationContent)) {
+      warnn('   MainApplication already extends NavigationApplication, skipping.')
       return applicationContent
-        .replace(/extends\s+Application\s+implements\s+ReactApplication/gi, "extends NavigationApplication")
-        .replace("import com.facebook.react.ReactApplication;", "import com.reactnativenavigation.NavigationApplication;");
     }
-    warnn("   MainApplication already extends NavigationApplication");
-    return applicationContent;
+
+    throw new Error('There was a problem extending NavigationApplication from your MainApplication file.')
   }
 
   _doesExtendApplication(applicationContent) {
-    return /\s+MainApplication\s+extends\s+Application\s+implements\s+ReactApplication\s+/.test(applicationContent);
+    return /\s+MainApplication\s+extends\s+Application\s+implements\s+ReactApplication\s+/.test(applicationContent)
+  }
+
+  _hasAlreadyLinkedApplication(applicationContent) {
+    return /\s+extends\s+NavigationApplication\s+/.test(applicationContent)
   }
 
   _extendNavigationHost(applicationContent) {
     if (this._doesExtendReactNativeHost(applicationContent)) {
-      debugn("   Changing host implementation to NavigationReactNativeHost");
+      debugn('   Changing host implementation to NavigationReactNativeHost')
+      return applicationContent
+        .replace('new ReactNativeHost(this)', 'new NavigationReactNativeHost(this)')
+        .replace(
+          'import com.facebook.react.ReactNativeHost;',
+          'import com.facebook.react.ReactNativeHost;\nimport com.reactnativenavigation.react.NavigationReactNativeHost;'
```

Comment:
You're keeping `import com.facebook.react.ReactNativeHost;`
Is that import needed?

Reply:
I will check, I tried to leave most of the logic unchanged, aside from things that I saw were clearly wrong or were unused code.

Reply:
It is indeed needed, but Android Studio points out that the import of `import android.app.Application;` and `import com.facebook.soloader.SoLoader;` are not necessary after the script runs, perhaps we could add this in a separate PR?